File: app/test/sidecar_supervisor_test.dart
import 'dart:async';
import 'dart:convert';

import 'package:auto_scoring_app/api/sidecar_api_client.dart';
import 'package:auto_scoring_app/core/sidecar_supervisor.dart';
import 'package:flutter_test/flutter_test.dart';

/// Covers every state transition of the sidecar supervisor against a fake
/// [SidecarPlatform]: startup, the handshake write race, health polling,
/// startup timeout, immediate exit, "already running", crash recovery,
/// restart, and shutdown.
///
/// Nothing here touches a process, a file, a socket, or the wall clock, so
/// the whole Windows-targeted lifecycle is verified on a Linux CI runner --
/// which is the point (`AGENTS.md`: inject the boundaries).
void main() {
  test('reaches ready once the handshake parses and health passes', () async {
    final platform = _FakePlatform();
    final supervisor = _supervisor(platform);
    addTearDown(supervisor.dispose);

    // The realistic ordering: the process is up for a while before it has
    // written anything, then the handshake appears, then it starts serving.
    platform.onPoll = (poll) {
      if (poll == 2) platform.writeHandshake(port: 51234, token: 'tok');
      if (poll >= 4) platform.healthy = true;
    };

    await supervisor.start();

    final state = supervisor.state.value;
    expect(state, isA<SidecarReady>());
    expect(
      (state as SidecarReady).connection.baseUrl,
      'http://127.0.0.1:51234',
    );
    expect(state.connection.token, 'tok');
  });

  test('passes --port 0 and the handshake path it created', () async {
    final platform = _FakePlatform()..healthy = true;
    platform.onPoll = (_) => platform.writeHandshake(port: 1, token: 't');
    final supervisor = _supervisor(platform);
    addTearDown(supervisor.dispose);

    await supervisor.start();

    expect(platform.spawns.single.executable, '/opt/sidecar');
    expect(platform.spawns.single.arguments, [
      '--port',
      '0',
      '--handshake-file',
      platform.handshakePaths.single,
    ]);
  });

  test(
    'omits --app-data-dir unless one was configured, and passes it when it was',
    () async {
      final platform = _FakePlatform()..healthy = true;
      platform.onPoll = (_) => platform.writeHandshake(port: 1, token: 't');
      final supervisor = SidecarSupervisor(
        platform: platform,
        executablePath: '/opt/sidecar',
        appDataDirectory: r'D:\data',
      );
      addTearDown(supervisor.dispose);

      await supervisor.start();

      expect(
        platform.spawns.single.arguments,
        containsAllInOrder(['--app-data-dir', r'D:\data']),
      );
    },
  );

  group('handshake reading', () {
    // The race the sidecar's own writer creates: `createHandshakeFile` makes
    // the file, and a *different* process fills it in later. Anything short
    // of a complete, parseable object means "not yet", never "give up".
    for (final (name, contents) in [
      ('an empty file', ''),
      ('whitespace only', '   \n'),
      ('a truncated write', '{"host": "127.0.0.1", "po'),
      ('valid JSON that is not an object', '["127.0.0.1", 1234]'),
      ('a missing token', '{"host": "127.0.0.1", "port": 1234}'),
      ('an empty token', '{"host": "127.0.0.1", "port": 1, "token": ""}'),
      (
        'a port that is not a number',
        '{"host": "127.0.0.1", "port": "1234", "token": "t"}',
      ),
    ]) {
      test('retries past $name', () async {
        final platform = _FakePlatform()..handshakeContents = contents;
        final supervisor = _supervisor(platform);
        addTearDown(supervisor.dispose);

        platform.onPoll = (poll) {
          if (poll == 3) {
            platform.writeHandshake(port: 4242, token: 'good');
            platform.healthy = true;
          }
        };

        await supervisor.start();

        expect(supervisor.state.value, isA<SidecarReady>());
        expect(
          (supervisor.state.value as SidecarReady).connection.baseUrl,
          'http://127.0.0.1:4242',
        );
      });
    }
  });

  test('deletes the handshake file once the token has been read', () async {
    final platform = _FakePlatform()..healthy = true;
    platform.onPoll = (_) => platform.writeHandshake(port: 1, token: 't');
    final supervisor = _supervisor(platform);
    addTearDown(supervisor.dispose);

    await supervisor.start();

    expect(supervisor.state.value, isA<SidecarReady>());
    expect(platform.deletedHandshakePaths, platform.handshakePaths);
  });

  test(
    'fails without spawning anything when no executable was found',
    () async {
      final platform = _FakePlatform();
      final supervisor = SidecarSupervisor(
        platform: platform,
        executablePath: null,
      );
      addTearDown(supervisor.dispose);

      await supervisor.start();

      expect(
        supervisor.state.value,
        isA<SidecarFailed>().having(
          (s) => s.failure,
          'failure',
          SidecarFailure.executableMissing,
        ),
      );
      expect(platform.spawns, isEmpty);
    },
  );

  test('reports a spawn that throws as a missing executable', () async {
    final platform = _FakePlatform()..spawnError = FakeSpawnFailure();
    final supervisor = _supervisor(platform);
    addTearDown(supervisor.dispose);

    await supervisor.start();

    expect(
      supervisor.state.value,
      isA<SidecarFailed>().having(
        (s) => s.failure,
        'failure',
        SidecarFailure.executableMissing,
      ),
    );
    // Even a failed spawn must not leave the temp handshake directory behind.
    expect(platform.deletedHandshakePaths, platform.handshakePaths);
  });

  test('reports an immediate exit without waiting out the timeout', () async {
    final platform = _FakePlatform();
    final supervisor = _supervisor(platform);
    addTearDown(supervisor.dispose);

    platform.onPoll = (poll) {
      if (poll == 1) platform.exitProcess(1);
    };

    await supervisor.start();

    expect(
      supervisor.state.value,
      isA<SidecarFailed>()
          .having(
            (s) => s.failure,
            'failure',
            SidecarFailure.exitedDuringStartup,
          )
          .having((s) => s.exitCode, 'exitCode', 1),
    );
    // The whole point: it gave up in a couple of polls, not 60 seconds of
    // them.
    expect(platform.elapsed, lessThan(const Duration(seconds: 5)));
  });

  test(
    'reports the sidecar refusing a second instance as alreadyRunning',
    () async {
      final platform = _FakePlatform();
      final supervisor = _supervisor(platform);
      addTearDown(supervisor.dispose);

      platform.onPoll = (poll) {
        if (poll == 1) platform.exitProcess(sidecarAlreadyRunningExitCode);
      };

      await supervisor.start();

      expect(
        supervisor.state.value,
        isA<SidecarFailed>().having(
          (s) => s.failure,
          'failure',
          SidecarFailure.alreadyRunning,
        ),
      );
    },
  );

  test('kills a sidecar that never becomes healthy, and times out', () async {
    final platform = _FakePlatform(); // never healthy, never exits
    final supervisor = _supervisor(platform);
    addTearDown(supervisor.dispose);
    platform.onPoll = (_) => platform.writeHandshake(port: 1, token: 't');

    await supervisor.start();

    expect(
      supervisor.state.value,
      isA<SidecarFailed>().having(
        (s) => s.failure,
        'failure',
        SidecarFailure.startupTimedOut,
      ),
    );
    expect(platform.spawns.single.killed, isTrue);
    expect(platform.elapsed, greaterThanOrEqualTo(sidecarStartupTimeout));
    expect(platform.deletedHandshakePaths, platform.handshakePaths);
  });

  test('reports an exit after ready as a crash', () async {
    final platform = _FakePlatform()..healthy = true;
    platform.onPoll = (_) => platform.writeHandshake(port: 1, token: 't');
    final supervisor = _supervisor(platform);
    addTearDown(supervisor.dispose);

    await supervisor.start();
    expect(supervisor.state.value, isA<SidecarReady>());

    platform.exitProcess(-1);
    await pumpEventQueue();

    expect(
      supervisor.state.value,
      isA<SidecarFailed>()
          .having((s) => s.failure, 'failure', SidecarFailure.crashed)
          .having((s) => s.exitCode, 'exitCode', -1),
    );
  });

  test(
    'restarting after a crash spawns a fresh sidecar and recovers',
    () async {
      final platform = _FakePlatform()..healthy = true;
      platform.onPoll = (_) => platform.writeHandshake(port: 1, token: 'first');
      final supervisor = _supervisor(platform);
      addTearDown(supervisor.dispose);

      await supervisor.start();
      platform.exitProcess(-1);
      await pumpEventQueue();
      expect(supervisor.state.value, isA<SidecarFailed>());

      // The 再起動 button (simplified-design-specification.md §24).
      platform.onPoll = (_) =>
          platform.writeHandshake(port: 2, token: 'second');
      await supervisor.start();

      expect(supervisor.state.value, isA<SidecarReady>());
      expect(
        (supervisor.state.value as SidecarReady).connection.token,
        'second',
      );
      expect(platform.spawns, hasLength(2));
      // Each attempt gets its own handshake file, so a previous attempt's
      // contents can never be read as this one's.
      expect(platform.handshakePaths.toSet(), hasLength(2));
    },
  );

  test(
    'a restart kills the process the previous attempt left running',
    () async {
      final platform = _FakePlatform()..healthy = true;
      platform.onPoll = (_) => platform.writeHandshake(port: 1, token: 't');
      final supervisor = _supervisor(platform);
      addTearDown(supervisor.dispose);

      await supervisor.start();
      await supervisor.start();

      expect(platform.spawns, hasLength(2));
      expect(platform.spawns.first.killed, isTrue);
      expect(platform.spawns.last.killed, isFalse);
    },
  );

  test('shutdown kills the sidecar and stops reporting crashes', () async {
    final platform = _FakePlatform()..healthy = true;
    platform.onPoll = (_) => platform.writeHandshake(port: 1, token: 't');
    final supervisor = _supervisor(platform);
    addTearDown(supervisor.dispose);

    await supervisor.start();
    await supervisor.shutdown();

    expect(supervisor.state.value, isA<SidecarStopped>());
    expect(platform.spawns.single.killed, isTrue);

    // The kill itself completes the exitCode future; a deliberate shutdown
    // must not then be re-reported to the UI as a crash.
    await pumpEventQueue();
    expect(supervisor.state.value, isA<SidecarStopped>());
  });

  test('the exit code shared with the sidecar has not drifted', () {
    // `backend/src/auto_scoring/api/sidecar.py::ALREADY_RUNNING_EXIT_CODE`.
    // The two constants are the whole protocol behind the "already running"
    // error screen, and nothing else would catch them diverging.
    expect(sidecarAlreadyRunningExitCode, 3);
  });
}

SidecarSupervisor _supervisor(_FakePlatform platform) =>
    SidecarSupervisor(platform: platform, executablePath: '/opt/sidecar');

/// Stands in for `Process.start` failing on a path that is not executable.
class FakeSpawnFailure implements Exception {}

class _Spawn {
  _Spawn(this.executable, this.arguments);

  final String executable;
  final List<String> arguments;
  final Completer<int> exit = Completer<int>();
  bool killed = false;
}

class _FakeProcess implements SidecarProcessHandle {
  _FakeProcess(this._spawn);

  final _Spawn _spawn;

  @override
  Future<int> get exitCode => _spawn.exit.future;

  @override
  void kill() {
    _spawn.killed = true;
    if (!_spawn.exit.isCompleted) _spawn.exit.complete(-9);
  }
}

/// An operating system that never sleeps: [delay] advances a counter instead
/// of real time, so a 60-second startup timeout is exercised instantly.
class _FakePlatform implements SidecarPlatform {
  final List<_Spawn> spawns = [];
  final List<String> handshakePaths = [];
  final List<String> deletedHandshakePaths = [];

  /// Contents the handshake file currently has. `null` means "no file yet".
  String? handshakeContents;

  bool healthy = false;
  Object? spawnError;

  /// Called before each poll's `delay`, with the 1-based poll number, so a
  /// test can describe *when* the sidecar reaches each milestone.
  void Function(int poll)? onPoll;

  Duration elapsed = Duration.zero;
  int _polls = 0;
  int _handshakeCounter = 0;

  void writeHandshake({required int port, required String token}) {
    handshakeContents = jsonEncode({
      'host': '127.0.0.1',
      'port': port,
      'token': token,
    });
  }

  void exitProcess(int code) {
    final spawn = spawns.last;
    if (!spawn.exit.isCompleted) spawn.exit.complete(code);
  }

  @override
  Future<String> createHandshakeFile() async {
    handshakeContents = null;
    final path = '/tmp/handshake-${_handshakeCounter++}.json';
    handshakePaths.add(path);
    return path;
  }

  @override
  Future<String?> readHandshakeFile(String path) async => handshakeContents;

  @override
  Future<void> deleteHandshakeFile(String path) async {
    deletedHandshakePaths.add(path);
  }

  @override
  Future<SidecarProcessHandle> start(
    String executable,
    List<String> arguments,
  ) async {
    if (spawnError case final error?) throw error;
    final spawn = _Spawn(executable, arguments);
    spawns.add(spawn);
    return _FakeProcess(spawn);
  }

  @override
  Future<bool> probeHealth(SidecarConnection connection) async => healthy;

  @override
  Future<void> delay(Duration duration) async {
    elapsed += duration;
    onPoll?.call(++_polls);
    // Yields to the event loop so a completed `exitCode` future's listener
    // runs, exactly as a real delay would.
    await Future<void>.delayed(Duration.zero);
  }

  @override
  DateTime now() => DateTime.utc(2026, 1, 1).add(elapsed);
}

File: desktop/test/sidecar-supervisor.test.ts
import { describe, expect, it } from "vitest";
import {
  SIDECAR_ALREADY_RUNNING_EXIT_CODE,
  SIDECAR_STARTUP_TIMEOUT_MS,
  SidecarSupervisor,
} from "../src/main/sidecar-supervisor";
import type {
  SidecarPlatform,
  SidecarProcessHandle,
} from "../src/main/sidecar-platform";

interface SpawnCall {
  readonly executable: string;
  readonly args: string[];
}

class FakeProcessHandle implements SidecarProcessHandle {
  readonly pid = 99999;
  private _resolveExit!: (code: number | null) => void;
  readonly exitCode = new Promise<number | null>((resolve) => {
    this._resolveExit = resolve;
  });

  killed = false;
  killSignal: string | null = null;

  kill(signal: NodeJS.Signals = "SIGTERM"): void {
    this.killed = true;
    this.killSignal = signal;
    this._resolveExit(-1);
  }

  exitProcess(code: number): void {
    this._resolveExit(code);
  }
}

class FakeSidecarPlatform implements SidecarPlatform {
  currentTime = 1_000_000;
  healthy = false;
  pollCount = 0;
  onPoll?: (poll: number) => void;

  createdDirs: string[] = [];
  deletedDirs: string[] = [];
  handshakeContents: string | null = null;
  spawns: SpawnCall[] = [];
  spawnError: Error | null = null;
  lastHandle?: FakeProcessHandle;

  async createHandshakeDirectory(): Promise<{
    dirPath: string;
    filePath: string;
  }> {
    const dirPath = `/tmp/fake-handshake-${this.createdDirs.length}`;
    this.createdDirs.push(dirPath);
    return { dirPath, filePath: `${dirPath}/handshake.json` };
  }

  async readHandshakeFile(_filePath: string): Promise<string | null> {
    return this.handshakeContents;
  }

  async deleteHandshakeDirectory(dirPath: string): Promise<void> {
    this.deletedDirs.push(dirPath);
  }

  async start(
    executable: string,
    args: string[],
  ): Promise<SidecarProcessHandle> {
    if (this.spawnError) {
      throw this.spawnError;
    }
    this.spawns.push({ executable, args });
    const handle = new FakeProcessHandle();
    this.lastHandle = handle;
    return handle;
  }

  async probeHealth(_baseUrl: string): Promise<boolean> {
    return this.healthy;
  }

  async delay(ms: number): Promise<void> {
    this.currentTime += ms;
    this.pollCount++;
    this.onPoll?.(this.pollCount);
  }

  now(): number {
    return this.currentTime;
  }

  currentPid(): number {
    return 12345;
  }

  writeHandshake(port = 51234, token = "test-token-value"): void {
    this.handshakeContents = JSON.stringify({
      host: "127.0.0.1",
      port,
      token,
    });
  }
}

describe("SidecarSupervisor unit tests", () => {
  it("reaches ready once the handshake parses and health check passes (INV-030)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.onPoll = (poll) => {
      if (poll === 2) {
        platform.writeHandshake(51234, "secret-token-abc");
      }
      if (poll >= 4) {
        platform.healthy = true;
      }
    };

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/opt/sidecar",
    });

    await supervisor.start();

    expect(supervisor.status.kind).toBe("ready");
    if (supervisor.status.kind === "ready") {
      expect(supervisor.status.connection.host).toBe("127.0.0.1");
      expect(supervisor.status.connection.port).toBe(51234);
      expect(supervisor.status.connection).not.toHaveProperty("token");
    }
    if (supervisor.internalStatus.kind === "ready") {
      expect(supervisor.internalStatus.connection.token).toBe(
        "secret-token-abc",
      );
    }
  });

  it("passes --port 0, --handshake-file, and --parent-pid (UG-01)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.healthy = true;
    platform.onPoll = () => platform.writeHandshake();

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/opt/sidecar",
    });

    await supervisor.start();

    expect(platform.spawns.length).toBe(1);
    const spawnCall = platform.spawns[0]!;
    expect(spawnCall.executable).toBe("/opt/sidecar");
    expect(spawnCall.args).toContain("--port");
    expect(spawnCall.args).toContain("0");
    expect(spawnCall.args).toContain("--handshake-file");
    expect(spawnCall.args).toContain("--parent-pid");
    expect(spawnCall.args).toContain("12345");
  });

  it("omits --app-data-dir unless configured, and passes it when configured (UG-06)", async () => {
    const platform1 = new FakeSidecarPlatform();
    platform1.healthy = true;
    platform1.onPoll = () => platform1.writeHandshake();

    const supervisorDefault = new SidecarSupervisor({
      platform: platform1,
      executablePath: "/opt/sidecar",
    });
    await supervisorDefault.start();
    expect(platform1.spawns[0]!.args).not.toContain("--app-data-dir");

    const platform2 = new FakeSidecarPlatform();
    platform2.healthy = true;
    platform2.onPoll = () => platform2.writeHandshake();

    const supervisorCustom = new SidecarSupervisor({
      platform: platform2,
      executablePath: "/opt/sidecar",
      appDataDirectory: "/custom/data",
    });
    await supervisorCustom.start();
    expect(platform2.spawns[0]!.args).toContain("--app-data-dir");
    expect(platform2.spawns[0]!.args).toContain("/custom/data");
  });

  describe("handshake reading retries (INV-037)", () => {
    const invalidPayloads = [
      ["empty file", ""],
      ["whitespace only", "   \n\t  "],
      ["truncated JSON", '{"host": "127.0.0.1", "po'],
      ["JSON array instead of object", '["127.0.0.1", 1234]'],
      ["missing token", '{"host": "127.0.0.1", "port": 1234}'],
      ["empty token", '{"host": "127.0.0.1", "port": 1234, "token": ""}'],
      [
        "port not a number",
        '{"host": "127.0.0.1", "port": "1234", "token": "t"}',
      ],
    ] as const;

    for (const [description, invalidContent] of invalidPayloads) {
      it(`retries past ${description}`, async () => {
        const platform = new FakeSidecarPlatform();
        platform.handshakeContents = invalidContent;

        platform.onPoll = (poll) => {
          if (poll === 3) {
            platform.writeHandshake(4321, "valid-token");
            platform.healthy = true;
          }
        };

        const supervisor = new SidecarSupervisor({
          platform,
          executablePath: "/opt/sidecar",
        });

        await supervisor.start();
        expect(supervisor.status.kind).toBe("ready");
        if (supervisor.internalStatus.kind === "ready") {
          expect(supervisor.internalStatus.connection.port).toBe(4321);
          expect(supervisor.internalStatus.connection.token).toBe(
            "valid-token",
          );
        }
      });
    }
  });

  it("deletes the handshake directory after token is read (INV-038)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.healthy = true;
    platform.onPoll = () => platform.writeHandshake();

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/opt/sidecar",
    });

    await supervisor.start();

    expect(supervisor.status.kind).toBe("ready");
    expect(platform.deletedDirs).toEqual(platform.createdDirs);
  });

  it("fails without spawning when executable is null (INV-035)", async () => {
    const platform = new FakeSidecarPlatform();
    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: null,
    });

    await supervisor.start();

    expect(supervisor.status.kind).toBe("failed");
    if (supervisor.status.kind === "failed") {
      expect(supervisor.status.failure).toBe("executableMissing");
    }
    expect(platform.spawns.length).toBe(0);
  });

  it("reports a spawn error as executableMissing (INV-035)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.spawnError = new Error("spawn ENOENT");

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/bad/path",
    });

    await supervisor.start();

    expect(supervisor.status.kind).toBe("failed");
    if (supervisor.status.kind === "failed") {
      expect(supervisor.status.failure).toBe("executableMissing");
    }
    expect(platform.deletedDirs).toEqual(platform.createdDirs);
  });

  it("reports an immediate exit without waiting out the timeout (INV-039)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.onPoll = (poll) => {
      if (poll === 1) {
        platform.lastHandle?.exitProcess(1);
      }
    };

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/opt/sidecar",
    });

    const startTime = platform.currentTime;
    await supervisor.start();

    expect(supervisor.status.kind).toBe("failed");
    if (supervisor.status.kind === "failed") {
      expect(supervisor.status.failure).toBe("exitedDuringStartup");
      expect(supervisor.status.exitCode).toBe(1);
    }
    expect(platform.currentTime - startTime).toBeLessThan(5000);
  });

  it("reports exit code 3 as alreadyRunning (INV-034, INV-040)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.onPoll = (poll) => {
      if (poll === 1) {
        platform.lastHandle?.exitProcess(SIDECAR_ALREADY_RUNNING_EXIT_CODE);
      }
    };

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/opt/sidecar",
    });

    await supervisor.start();

    expect(supervisor.status.kind).toBe("failed");
    if (supervisor.status.kind === "failed") {
      expect(supervisor.status.failure).toBe("alreadyRunning");
      expect(supervisor.status.exitCode).toBe(3);
    }
  });

  it("times out and kills sidecar when health probe never passes (UG-08)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.onPoll = () => platform.writeHandshake(); // writes valid handshake but healthy is never true

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/opt/sidecar",
    });

    const startTime = platform.currentTime;
    await supervisor.start();

    expect(supervisor.status.kind).toBe("failed");
    if (supervisor.status.kind === "failed") {
      expect(supervisor.status.failure).toBe("startupTimedOut");
    }
    expect(platform.lastHandle?.killed).toBe(true);
    expect(platform.currentTime - startTime).toBeGreaterThanOrEqual(
      SIDECAR_STARTUP_TIMEOUT_MS,
    );
    expect(platform.deletedDirs).toEqual(platform.createdDirs);
  });

  it("reports exit after ready as crashed (INV-041)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.healthy = true;
    platform.onPoll = () => platform.writeHandshake();

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/opt/sidecar",
    });

    await supervisor.start();
    expect(supervisor.status.kind).toBe("ready");

    platform.lastHandle?.exitProcess(137);
    await new Promise((resolve) => setTimeout(resolve, 0));

    expect(supervisor.status.kind).toBe("failed");
    if (supervisor.status.kind === "failed") {
      expect(supervisor.status.failure).toBe("crashed");
      expect(supervisor.status.exitCode).toBe(137);
    }
  });

  it("shutdown kills the child and does not report crash (INV-031, INV-041)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.healthy = true;
    platform.onPoll = () => platform.writeHandshake();

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/opt/sidecar",
    });

    await supervisor.start();
    expect(supervisor.status.kind).toBe("ready");

    await supervisor.shutdown();
    expect(supervisor.status.kind).toBe("stopped");
    expect(platform.lastHandle?.killed).toBe(true);
  });

  it("restart terminates existing process and starts new attempt (INV-032)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.healthy = true;
    platform.onPoll = () => platform.writeHandshake(50001, "token-1");

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/opt/sidecar",
    });

    await supervisor.start();
    expect(supervisor.status.kind).toBe("ready");
    const firstHandle = platform.lastHandle!;

    platform.writeHandshake(50002, "token-2");
    await supervisor.restart();

    expect(firstHandle.killed).toBe(true);
    expect(supervisor.status.kind).toBe("ready");
    if (supervisor.internalStatus.kind === "ready") {
      expect(supervisor.internalStatus.connection.port).toBe(50002);
      expect(supervisor.internalStatus.connection.token).toBe("token-2");
    }
    expect(platform.spawns.length).toBe(2);
  });

  it("failed state never exposes bearer token (INV-036)", async () => {
    const platform = new FakeSidecarPlatform();
    platform.onPoll = (poll) => {
      if (poll === 1) {
        platform.writeHandshake(50000, "super-secret-token");
        platform.lastHandle?.exitProcess(1);
      }
    };

    const supervisor = new SidecarSupervisor({
      platform,
      executablePath: "/opt/sidecar",
    });

    await supervisor.start();

    expect(supervisor.status.kind).toBe("failed");
    const json = JSON.stringify(supervisor.status);
    expect(json).not.toContain("super-secret-token");
  });
});
